### DISCUSSION TITLE
Automatically unban users with more than a given number of files?

### DISCUSSION THREAD

**Discussion by ghost** [nicotine-plus/nicotine-plus#2102] [Q&A] [answered]
Can I remove the ban from users automatically when they increase the files above the minimum after they get banned by the Leech Detector?

**Comment by slook** [comment, marked as answer]
If you already banned a user then they will never again encounter your Leech Detector, therefore they will never get an opportunity to be unbanned!

If you wish to unban a user, call the unban_user() method:
`self.core.network_filter.unban_user(user)`

There is currently no way of tracking the reason why users are banned (it is a planned feature see #1763), and furthermore it might be difficult to obtain Stats for a banned user, if the NetworkFilter module may disallow response data from a `GetUserStats()` request, I'm not sure because I've never tried to do what you want to do.

Really the Leech Detector plugin should have a .JSON file to store users who are leechers so that it is persistent across sessions, primarily such that leechers are not messaged more than once (but then this stored data could be used for other purposes, or by other plugins, etc).

For such an implementation, it would indeed be necessary to periodically update the stored data for users whose number of files had increased since they were first identified as a leecher (but checking the Stats data for banned users might require extra effort by plugin developers, since automatic banning will never be a feature built in to the core plugin).

**Comment by ghost** [comment]
How can I do this at regular intervals or at startup if the list of leechers was saved on disk?

**Reply by slook** [reply]
> at regular intervals

On plugin load, create a Threading timer, then create a function with a While loop for doing what you need every X seconds, see transfers.py for an example implementation:
https://github.com/nicotine-plus/nicotine-plus/blob/4d30f1b8a2810c3292306755fed28c2c8ede9db7/pynicotine/transfers.py#L36
https://github.com/nicotine-plus/nicotine-plus/blob/4d30f1b8a2810c3292306755fed28c2c8ede9db7/pynicotine/transfers.py#L160-L164
https://github.com/nicotine-plus/nicotine-plus/blob/4d30f1b8a2810c3292306755fed28c2c8ede9db7/pynicotine/transfers.py#L2063-L2073

Avoid using a GLib timer because otherwise the plugin will crash in headless (no GUI) mode.

> at startup

Create some method(s) and call your code from `loaded_notification()`
